applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.cpp

```diff
@@ -0,0 +1,262 @@
+  
+        for(IndexType i = 0; i < num_dofs; ++i)
+        {
+             for(IndexType j = 0; j < num_GP; ++j)
+                rOutput(i, j) = length_derivative_vector[i] * derivative_pre_factor; 
+        }
+    }
+    else
+        KRATOS_ERROR << "Stress displacement derivative only available for Gauss-points quantities!" << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+double AdjointFiniteDifferenceTrussElement::GetPerturbationSizeModificationFactor(const Variable<array_1d<double,3>>& rDesignVariable)
+{
+    KRATOS_TRY;
+
+    if(rDesignVariable == SHAPE)
+        return this->CalculateReferenceLength();
+    else
+        return 1.0;
+
+    KRATOS_CATCH("")
+}
+
+double AdjointFiniteDifferenceTrussElement::CalculateReferenceLength() 
+{
+    KRATOS_TRY;
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const double dx = this->GetGeometry()[1].X0() - this->GetGeometry()[0].X0();
+    const double dy = this->GetGeometry()[1].Y0() - this->GetGeometry()[0].Y0();
+    const double dz = this->GetGeometry()[1].Z0() - this->GetGeometry()[0].Z0();
+    const double L = std::sqrt(dx * dx + dy * dy + dz * dz);
+
+    KRATOS_ERROR_IF(L<=numerical_limit)
+        << "Reference Length of element" << this->Id() << "~ 0" << std::endl;
+    return L;
+    KRATOS_CATCH("")
+}
+
+double AdjointFiniteDifferenceTrussElement::CalculateCurrentLength() 
+{
+    KRATOS_TRY;
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const double du =
+        this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_X) -
+        this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_X);
+    const double dv =
+        this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_Y) -
+        this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_Y);
+    const double dw =
+        this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_Z) -
+        this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_Z);
+    const double dx = this->GetGeometry()[1].X0() - this->GetGeometry()[0].X0();
+    const double dy = this->GetGeometry()[1].Y0() - this->GetGeometry()[0].Y0();
+    const double dz = this->GetGeometry()[1].Z0() - this->GetGeometry()[0].Z0();
+    const double l = std::sqrt((du + dx) * (du + dx) + (dv + dy) * (dv + dy) +
+                               (dw + dz) * (dw + dz));
+
+    KRATOS_ERROR_IF(l<=numerical_limit)
+        << "Current Length of element" << this->Id() << "~ 0" << std::endl;
+    return l;
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceTrussElement::CalculateCurrentLengthDisplacementDerivative(Vector& rDerivativeVector)
+{
+    KRATOS_TRY;
+
+    const SizeType number_of_nodes = this->GetGeometry().size();
+    const SizeType dimension = this->GetGeometry().WorkingSpaceDimension();
+    const SizeType num_dofs = number_of_nodes * dimension;
+
+    KRATOS_ERROR_IF_NOT(num_dofs == 6) << "This implementation expects 6 dofs!" << std::endl;
```

Comment:
This check should be covered in `Check`

Reply:
The dimension and the the number of nodes are already controlled in ```Check```. This error message is not necessary. I remove it.